### ISSUE TITLE
Unreal Engine 5.4 Support

### ISSUE THREAD

**Issue by OdinVex** [eddiebeazer/ue5-linter#10]
Any chance of a 5.4 update?

**Comment by eddiebeazer** [comment]
I plan to, just might not happen until July. I’m still on 5.3 and won’t need to upgrade until around that time. Mind sharing the compilation errors that you are getting?

**Comment by OdinVex** [comment]
I forgot to save the log. I compile on Linux anyway. There's nothing specific I've ever needed to do to get it working (as far as Linux goes).

**Comment by frankiezafe** [comment]
I did a quick fix for 5.4 (see pull request), i hope it might help you.
Contact me on twitter if you have troubles, i'm currently updating to 5.4 at the office.
@eddiebeazer : it's mainly missing inludes and updating the asset registry calls with TopLevelAssetPath instead of strings

**Comment by OdinVex** [comment]
> I did a quick fix for 5.4 (see pull request), i hope it might help you. Contact me on twitter if you have troubles, i'm currently updating to 5.4 at the office.

Unfortunately issues persist:
```
In file included from ../Plugins/Marketplace/Linter_5.4/Intermediate/Build/Linux/x64/UnrealEditor/Development/Linter/Module.Linter.cpp:33:
./Engine/Plugins/Marketplace/Linter_5.4/Source/Linter/Private/BatchRenameTool/BatchRenameTool.cpp:77:41: warning: 'AssetClass' is deprecated: Class names are now represented by path names. Please use AssetClassPath. Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile. [-Wdeprecated-declarations]
                        if (!(!Asset.IsRedirector() && Asset.AssetClass != NAME_Class && !(Asset.PackageFlags & PKG_FilterEditorOnly)))
                                                             ^
./Engine/Source/Runtime/CoreUObject/Public/AssetRegistry/AssetData.h:186:2: note: 'AssetClass' has been explicitly marked deprecated here
        UE_DEPRECATED(5.1, "Class names are now represented by path names. Please use AssetClassPath.")
        ^
./Engine/Source/Runtime/Core/Public/Misc/CoreMiscDefines.h:274:43: note: expanded from macro 'UE_DEPRECATED'
#define UE_DEPRECATED(Version, Message) [[deprecated(Message " Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile.")]]
                                          ^
In file included from ../Plugins/Marketplace/Linter_5.4/Intermediate/Build/Linux/x64/UnrealEditor/Development/Linter/Module.Linter.cpp:34:
./Engine/Plugins/Marketplace/Linter_5.4/Source/Linter/Private/Linter.cpp:122:33: error: no viable conversion from 'FName' to 'FTopLevelAssetPath'
        AssetRegistry.GetAssetsByClass(ULintRuleSet::StaticClass()->GetFName(), FoundRuleSets, true);
                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./Engine/Source/Runtime/CoreUObject/Public/UObject/TopLevelAssetPath.h:32:8: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'FName' to 'const FTopLevelAssetPath &' for 1st argument
struct FTopLevelAssetPath
       ^
./Engine/Source/Runtime/CoreUObject/Public/UObject/TopLevelAssetPath.h:32:8: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'FName' to 'FTopLevelAssetPath &&' for 1st argument
./Engine/Source/Runtime/CoreUObject/Public/UObject/TopLevelAssetPath.h:35:2: note: candidate constructor not viable: no known conversion from 'FName' to 'TYPE_OF_NULLPTR' (aka 'std::nullptr_t') for 1st argument
        FTopLevelAssetPath(TYPE_OF_NULLPTR) { }
        ^
./Engine/Source/Runtime/CoreUObject/Public/UObject/TopLevelAssetPath.h:46:11: note: explicit constructor is not a candidate
        explicit FTopLevelAssetPath(FName InPath)
                 ^
./Engine/Source/Runtime/CoreUObject/Public/UObject/TopLevelAssetPath.h:52:11: note: explicit constructor is not a candidate
        explicit FTopLevelAssetPath(const FString& Path) { TrySetPath(FStringView(Path)); }
                 ^
./Engine/Source/Runtime/CoreUObject/Public/UObject/TopLevelAssetPath.h:54:11: note: explicit constructor is not a candidate
        explicit FTopLevelAssetPath(TStringView<CharType> Path) { TrySetPath(Path); }
                 ^
./Engine/Source/Runtime/CoreUObject/Public/UObject/TopLevelAssetPath.h:56:11: note: explicit constructor is not a candidate
        explicit FTopLevelAssetPath(const CharType* Path) { TrySetPath(TStringView<CharType>(Path)); }
                 ^
./Engine/Source/Runtime/CoreUObject/Public/UObject/TopLevelAssetPath.h:59:11: note: explicit constructor is not a candidate
        explicit FTopLevelAssetPath(const UObject* InObject) { TrySetPath(InObject); }
                 ^
./Engine/Source/Runtime/AssetRegistry/Public/AssetRegistry/IAssetRegistry.h:248:51: note: passing argument to parameter 'ClassPathName' here
        virtual bool GetAssetsByClass(FTopLevelAssetPath ClassPathName, TArray<FAssetData>& OutAssetData, bool bSearchSubClasses = false) const = 0;
                                                         ^
In file included from ../Plugins/Marketplace/Linter_5.4/Intermediate/Build/Linux/x64/UnrealEditor/Development/Linter/Module.Linter.cpp:35:
./Engine/Plugins/Marketplace/Linter_5.4/Source/Linter/Private/LinterCommandlet.cpp:60:46: error: no viable conversion from 'FName' to 'FTopLevelAssetPath'
                AssetRegistryModule.Get().GetAssetsByClass(ULintRuleSet::StaticClass()->GetFName(), FoundRuleSets, true);
                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./Engine/Source/Runtime/CoreUObject/Public/UObject/TopLevelAssetPath.h:32:8: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'FName' to 'const FTopLevelAssetPath &' for 1st argument
struct FTopLevelAssetPath
       ^
./Engine/Source/Runtime/CoreUObject/Public/UObject/TopLevelAssetPath.h:32:8: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'FName' to 'FTopLevelAssetPath &&' for 1st argument
./Engine/Source/Runtime/CoreUObject/Public/UObject/TopLevelAssetPath.h:35:2: note: candidate constructor not viable: no known conversion from 'FName' to 'TYPE_OF_NULLPTR' (aka 'std::nullptr_t') for 1st argument
        FTopLevelAssetPath(TYPE_OF_NULLPTR) { }
        ^
./Engine/Source/Runtime/CoreUObject/Public/UObject/TopLevelAssetPath.h:46:11: note: explicit constructor is not a candidate
        explicit FTopLevelAssetPath(FName InPath)
                 ^
./Engine/Source/Runtime/CoreUObject/Public/UObject/TopLevelAssetPath.h:52:11: note: explicit constructor is not a candidate
        explicit FTopLevelAssetPath(const FString& Path) { TrySetPath(FStringView(Path)); }
                 ^
./Engine/Source/Runtime/CoreUObject/Public/UObject/TopLevelAssetPath.h:54:11: note: explicit constructor is not a candidate
        explicit FTopLevelAssetPath(TStringView<CharType> Path) { TrySetPath(Path); }
                 ^
./Engine/Source/Runtime/CoreUObject/Public/UObject/TopLevelAssetPath.h:56:11: note: explicit constructor is not a candidate
        explicit FTopLevelAssetPath(const CharType* Path) { TrySetPath(TStringView<CharType>(Path)); }
                 ^
./Engine/Source/Runtime/CoreUObject/Public/UObject/TopLevelAssetPath.h:59:11: note: explicit constructor is not a candidate
        explicit FTopLevelAssetPath(const UObject* InObject) { TrySetPath(InObject); }
                 ^
./Engine/Source/Runtime/AssetRegistry/Public/AssetRegistry/IAssetRegistry.h:248:51: note: passing argument to parameter 'ClassPathName' here
        virtual bool GetAssetsByClass(FTopLevelAssetPath ClassPathName, TArray<FAssetData>& OutAssetData, bool bSearchSubClasses = false) const = 0;
                                                         ^
In file included from ../Plugins/Marketplace/Linter_5.4/Intermediate/Build/Linux/x64/UnrealEditor/Development/Linter/Module.Linter.cpp:35:
./Engine/Plugins/Marketplace/Linter_5.4/Source/Linter/Private/LinterCommandlet.cpp:132:74: warning: 'ObjectPath' is deprecated: FName asset paths have been deprecated. Use GetSoftObjectPath to get the path this asset will use in memory when loaded or GetObjectPathString() if you were just doing ObjectPath.ToString() Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile. [-Wdeprecated-declarations]
                                AssetJsonObject->SetStringField(TEXT("ViolatorAssetPath"), AssetData.ObjectPath.ToString());
                                                                                                     ^
./Engine/Source/Runtime/CoreUObject/Public/AssetRegistry/AssetData.h:176:2: note: 'ObjectPath' has been explicitly marked deprecated here
        UE_DEPRECATED(5.1, "FName asset paths have been deprecated. Use GetSoftObjectPath to get the path this asset will use in memory when loaded or GetObjectPathString() if you were just doing ObjectPath.ToString()")
        ^
./Engine/Source/Runtime/Core/Public/Misc/CoreMiscDefines.h:274:43: note: expanded from macro 'UE_DEPRECATED'
#define UE_DEPRECATED(Version, Message) [[deprecated(Message " Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile.")]]
                                          ^
In file included from ../Plugins/Marketplace/Linter_5.4/Intermediate/Build/Linux/x64/UnrealEditor/Development/Linter/Module.Linter.cpp:35:
./Engine/Plugins/Marketplace/Linter_5.4/Source/Linter/Private/LinterCommandlet.cpp:265:72: error: '&&' within '||' [-Werror,-Wlogical-op-parentheses]
        if (NumErrors > 0 || Switches.Contains(TEXT("TreatWarningsAsErrors")) && NumWarnings > 0)
                          ~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
./Engine/Plugins/Marketplace/Linter_5.4/Source/Linter/Private/LinterCommandlet.cpp:265:72: note: place parentheses around the '&&' expression to silence this warning
        if (NumErrors > 0 || Switches.Contains(TEXT("TreatWarningsAsErrors")) && NumWarnings > 0)
                                                                              ^
                             (                                                                  )
In file included from ../Plugins/Marketplace/Linter_5.4/Intermediate/Build/Linux/x64/UnrealEditor/Development/Linter/Module.Linter.cpp:36:
./Engine/Plugins/Marketplace/Linter_5.4/Source/Linter/Private/LinterContentBrowserExtensions.cpp:89:41: warning: 'AssetClass' is deprecated: Class names are now represented by path names. Please use AssetClassPath. Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile. [-Wdeprecated-declarations]
                                        if (!Asset.IsRedirector() && Asset.AssetClass != NAME_Class && !(Asset.PackageFlags & PKG_FilterEditorOnly))
                                                                           ^
./Engine/Source/Runtime/CoreUObject/Public/AssetRegistry/AssetData.h:186:2: note: 'AssetClass' has been explicitly marked deprecated here
        UE_DEPRECATED(5.1, "Class names are now represented by path names. Please use AssetClassPath.")
        ^
./Engine/Source/Runtime/Core/Public/Misc/CoreMiscDefines.h:274:43: note: expanded from macro 'UE_DEPRECATED'
#define UE_DEPRECATED(Version, Message) [[deprecated(Message " Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile.")]]
                                          ^
In file included from ../Plugins/Marketplace/Linter_5.4/Intermediate/Build/Linux/x64/UnrealEditor/Development/Linter/Module.Linter.cpp:36:
./Engine/Plugins/Marketplace/Linter_5.4/Source/Linter/Private/LinterContentBrowserExtensions.cpp:126:41: warning: 'AssetClass' is deprecated: Class names are now represented by path names. Please use AssetClassPath. Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile. [-Wdeprecated-declarations]
                                        if (!Asset.IsRedirector() && Asset.AssetClass != NAME_Class && !(Asset.PackageFlags & PKG_FilterEditorOnly))
                                                                           ^
./Engine/Source/Runtime/CoreUObject/Public/AssetRegistry/AssetData.h:186:2: note: 'AssetClass' has been explicitly marked deprecated here
        UE_DEPRECATED(5.1, "Class names are now represented by path names. Please use AssetClassPath.")
        ^
./Engine/Source/Runtime/Core/Public/Misc/CoreMiscDefines.h:274:43: note: expanded from macro 'UE_DEPRECATED'
#define UE_DEPRECATED(Version, Message) [[deprecated(Message " Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile.")]]
                                          ^
In file included from ../Plugins/Marketplace/Linter_5.4/Intermediate/Build/Linux/x64/UnrealEditor/Development/Linter/Module.Linter.cpp:40:
./Engine/Plugins/Marketplace/Linter_5.4/Source/Linter/Private/LintRule.cpp:242:37: warning: 'GetAssetByObjectPath' is deprecated: Asset path FNames have been deprecated, use Soft Object Path instead. Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile. [-Wdeprecated-declarations]
                ViolatorAssetData = AssetRegistry.GetAssetByObjectPath(FName(*Violator->GetPathName()));
                                                  ^
./Engine/Source/Runtime/AssetRegistry/Public/AssetRegistry/IAssetRegistry.h:300:2: note: 'GetAssetByObjectPath' has been explicitly marked deprecated here
        UE_DEPRECATED(5.1, "Asset path FNames have been deprecated, use Soft Object Path instead.")
        ^
./Engine/Source/Runtime/Core/Public/Misc/CoreMiscDefines.h:274:43: note: expanded from macro 'UE_DEPRECATED'
#define UE_DEPRECATED(Version, Message) [[deprecated(Message " Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile.")]]
                                          ^
In file included from ../Plugins/Marketplace/Linter_5.4/Intermediate/Build/Linux/x64/UnrealEditor/Development/Linter/Module.Linter.cpp:64:
./Engine/Plugins/Marketplace/Linter_5.4/Source/Linter/Private/LintRuleSet.cpp:85:99: warning: 'ObjectPath' is deprecated: FName asset paths have been deprecated. Use GetSoftObjectPath to get the path this asset will use in memory when loaded or GetObjectPathString() if you were just doing ObjectPath.ToString() Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile. [-Wdeprecated-declarations]
                        Threads.Push(FRunnableThread::Create(Runner, *FString::Printf(TEXT("FLintRunner - %s"), *Asset.ObjectPath.ToString()), 0, TPri_Normal));
                                                                                                                       ^
./Engine/Source/Runtime/CoreUObject/Public/AssetRegistry/AssetData.h:176:2: note: 'ObjectPath' has been explicitly marked deprecated here
        UE_DEPRECATED(5.1, "FName asset paths have been deprecated. Use GetSoftObjectPath to get the path this asset will use in memory when loaded or GetObjectPathString() if you were just doing ObjectPath.ToString()")
        ^
./Engine/Source/Runtime/Core/Public/Misc/CoreMiscDefines.h:274:43: note: expanded from macro 'UE_DEPRECATED'
#define UE_DEPRECATED(Version, Message) [[deprecated(Message " Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile.")]]
                                          ^
In file included from ../Plugins/Marketplace/Linter_5.4/Intermediate/Build/Linux/x64/UnrealEditor/Development/Linter/Module.Linter.cpp:68:
./Engine/Plugins/Marketplace/Linter_5.4/Source/Linter/Private/UI/LintReport.cpp:285:73: warning: 'ObjectPath' is deprecated: FName asset paths have been deprecated. Use GetSoftObjectPath to get the path this asset will use in memory when loaded or GetObjectPathString() if you were just doing ObjectPath.ToString() Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile. [-Wdeprecated-declarations]
                        AssetJsonObject->SetStringField(TEXT("ViolatorAssetPath"), AssetData.ObjectPath.ToString());
                                                                                             ^
./Engine/Source/Runtime/CoreUObject/Public/AssetRegistry/AssetData.h:176:2: note: 'ObjectPath' has been explicitly marked deprecated here
        UE_DEPRECATED(5.1, "FName asset paths have been deprecated. Use GetSoftObjectPath to get the path this asset will use in memory when loaded or GetObjectPathString() if you were just doing ObjectPath.ToString()")
        ^
./Engine/Source/Runtime/Core/Public/Misc/CoreMiscDefines.h:274:43: note: expanded from macro 'UE_DEPRECATED'
#define UE_DEPRECATED(Version, Message) [[deprecated(Message " Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile.")]]
                                          ^
In file included from ../Plugins/Marketplace/Linter_5.4/Intermediate/Build/Linux/x64/UnrealEditor/Development/Linter/Module.Linter.cpp:72:
./Engine/Plugins/Marketplace/Linter_5.4/Source/Linter/Private/UI/LintReportRuleDetails.cpp:62:32: warning: 'GetAssetByObjectPath' is deprecated: Asset path FNames have been deprecated, use Soft Object Path instead. Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile. [-Wdeprecated-declarations]
        RuleAssetData = AssetRegistry.GetAssetByObjectPath(FName(*BrokenRule->GetPathName()), true);
                                      ^
./Engine/Source/Runtime/AssetRegistry/Public/AssetRegistry/IAssetRegistry.h:300:2: note: 'GetAssetByObjectPath' has been explicitly marked deprecated here
        UE_DEPRECATED(5.1, "Asset path FNames have been deprecated, use Soft Object Path instead.")
        ^
./Engine/Source/Runtime/Core/Public/Misc/CoreMiscDefines.h:274:43: note: expanded from macro 'UE_DEPRECATED'
#define UE_DEPRECATED(Version, Message) [[deprecated(Message " Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile.")]]
                                          ^
In file included from ../Plugins/Marketplace/Linter_5.4/Intermediate/Build/Linux/x64/UnrealEditor/Development/Linter/Module.Linter.cpp:73:
./Engine/Plugins/Marketplace/Linter_5.4/Source/Linter/Private/UI/LintReportRuleError.cpp:47:108: warning: 'ObjectPath' is deprecated: FName asset paths have been deprecated. Use GetSoftObjectPath to get the path this asset will use in memory when loaded or GetObjectPathString() if you were just doing ObjectPath.ToString() Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile. [-Wdeprecated-declarations]
                                        AssetDatas.Push(AssetRegistryModule.Get().GetAssetByObjectPath(RuleViolation.Get()->ViolatorAssetData.ObjectPath));
                                                                                                                                              ^
./Engine/Source/Runtime/CoreUObject/Public/AssetRegistry/AssetData.h:176:2: note: 'ObjectPath' has been explicitly marked deprecated here
        UE_DEPRECATED(5.1, "FName asset paths have been deprecated. Use GetSoftObjectPath to get the path this asset will use in memory when loaded or GetObjectPathString() if you were just doing ObjectPath.ToString()")
        ^
./Engine/Source/Runtime/Core/Public/Misc/CoreMiscDefines.h:274:43: note: expanded from macro 'UE_DEPRECATED'
#define UE_DEPRECATED(Version, Message) [[deprecated(Message " Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile.")]]
                                          ^
In file included from ../Plugins/Marketplace/Linter_5.4/Intermediate/Build/Linux/x64/UnrealEditor/Development/Linter/Module.Linter.cpp:73:
./Engine/Plugins/Marketplace/Linter_5.4/Source/Linter/Private/UI/LintReportRuleError.cpp:47:48: warning: 'GetAssetByObjectPath' is deprecated: Asset path FNames have been deprecated, use Soft Object Path instead. Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile. [-Wdeprecated-declarations]
                                        AssetDatas.Push(AssetRegistryModule.Get().GetAssetByObjectPath(RuleViolation.Get()->ViolatorAssetData.ObjectPath));
                                                                                  ^
./Engine/Source/Runtime/AssetRegistry/Public/AssetRegistry/IAssetRegistry.h:300:2: note: 'GetAssetByObjectPath' has been explicitly marked deprecated here
        UE_DEPRECATED(5.1, "Asset path FNames have been deprecated, use Soft Object Path instead.")
        ^
./Engine/Source/Runtime/Core/Public/Misc/CoreMiscDefines.h:274:43: note: expanded from macro 'UE_DEPRECATED'
#define UE_DEPRECATED(Version, Message) [[deprecated(Message " Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile.")]]
                                          ^
In file included from ../Plugins/Marketplace/Linter_5.4/Intermediate/Build/Linux/x64/UnrealEditor/Development/Linter/Module.Linter.cpp:74:
./Engine/Plugins/Marketplace/Linter_5.4/Source/Linter/Private/UI/LintReportRuleErrorList.cpp:3:10: fatal error: 'UI/LintReportruleErrorList.h' file not found
```

**Comment by eddiebeazer** [comment]
Closed by #12

**Comment by OdinVex** [comment]
First issue: `if (NumErrors > 0 || Switches.Contains(TEXT("TreatWarningsAsErrors")) && NumWarnings > 0)` in `Source/Linter/Private/` is still ambiguous (`error: '&&' within '||' [-Werror,-Wlogical-op-parentheses]`). Enclose everything after `||` with parenthesis.

Second issue: `Source/Linter/Private/UI/LintReportRuleErrorList.cpp:3:10: fatal error: 'UI/LintReportruleErrorList.h' file not found #include "UI/LintReportruleErrorList.h"` (Notice the case, it needs to be 'Rule'.)

Third issue: `Source/Linter/Private/UI/LintWizard.cpp:31:9: error: 'LOCTEXT_NAMESPACE' macro redefined [-Werror,-Wmacro-redefined]` -> `Source/Linter/Private/UI/LintReportRuleErrorList.cpp:23:9: note: previous definition is here` It's because `#undef LOCTEXT_NAMESPACE` was not at the end of `Source/Linter/Private/UI/LintReportRuleErrorList.cpp`.
